packages/flutter/lib/src/material/colors.dart

```diff
@@ -88,6 +90,133 @@ class MaterialAccentColor extends ColorSwatch<int> {
+/// based on a given set of states.
+typedef ColorResolver = Color Function(Set<MaterialState> states);
+
+/// Defines a [Color] that can react to changes in the state of a Material
+/// component, based on a given set of [MaterialState]s.
+///
+/// This is useful for preserving the accessibility of text in different states
+/// of a component. For example, in a [FlatButton] with blue text, the text will
+/// become more difficult to read when the button is hovered, focused, or pressed,
+/// because the contrast ratio between the button and the text will decrease. To
+/// solve this, you can use [MaterialStateColor] to make the text darker when the
+/// [FlatButton] is hovered, focused, or pressed.
+///
+/// To use a [MaterialStateColor], you can either:
+///   1. Create a subclass of [MaterialStateColor] and implement the abstract `resolve` method.
+///   2. Use [MaterialStateColor.resolveWith] and pass in a callback that
+///      will be used to resolve the color in the given states.
+///
+/// This should only be used as parameters when they are documented to take
+/// [MaterialStatefulColor], otherwise only the default state will be used.
+///
+/// Example:
+/// ```dart
+/// Color getTextColor(Set<MaterialSet> states) {
+///   final interactiveStates = <MaterialState>{
+///     MaterialState.pressed,
+///     MaterialState.hovered,
+///     MaterialState.focused,
+///   };
+///   if(states.any(interactiveStates.contains) {
+///     return Colors.blue[900];
+///   }
+///   return Colors.blue[600];
+/// }
+///
+/// FlatButton(
+///   ...
+///   textColor: MaterialStateColor.resolveWith(getTextColor),
+/// ),
+/// ```
+abstract class MaterialStateColor extends Color {
+  /// Creates a [MaterialStateColor].
+  ///
+  /// If you want a `const` [MaterialStateColor], you'll need to extend
+  /// [MaterialStateColor] and override the [resolve] method. You'll also need
+  /// to provide a `defaultValue` to the super constructor, so that we can know
+  /// at compile-time what the value of the default [Color] is.
+  ///
+  /// Example (`const`):
+  ///
+  /// ```dart
+  /// class TextColor extends MaterialStateColor {
+  ///   static const int _defaultColor = 0xcafefeed;
+  ///   static const int _pressedColor = 0xdeadbeef;
+  ///
+  ///   const TextColor() : super(_defaultColor);
+  ///
+  ///   @override
+  ///   Color resolve(Set<MaterialState> states) {
+  ///     if (states.contains(MaterialState.pressed)) {
+  ///       return const Color(_pressedColor);
+  ///     }
+  ///     return const Color(_defaultColor);
+  ///   }
+  /// }
+  /// ```
+  ///
+  /// If you don't need your [MaterialStateColor] to be `const`, consider using
+  /// [MaterialStateColor.resolveWith].
+  const MaterialStateColor(int defaultValue) : super(defaultValue);
+
+  /// Creates a [MaterialStateColor] from a [ColorResolver] callback function.
+  ///
+  /// If used as a regular color, the color resolved in the default state will
```

Comment:
Might be helpful to clarify that the "default state" is the empty set, here, rather than in the next paragraph.

Reply:
Done.